### PR TITLE
short scale formatter

### LINKED ISSUE #1353: Add short-scale formatter
http://en.wikipedia.org/wiki/Long_and_short_scales

The current formatter for large numbers, `Formatters.siSuffix`, produces some unintuitive output for small numbers (0.1 --> "100m") and large numbers (10^9 --> "1.00G") for people who are used to the short scale. Here's probably what would be expected:

* 0.1 --> "0.1" (maybe the same behavior as the fixed formatter)
* 1 000 000 --> "1 M" (one million)
* 1 000 000 000 --> "1 B" (one billion)
* 1 000 000 000 000 --> "1 T" (one trillion)

Not sure if we should go above this to "quadrillion"

### PR DESCRIPTION
**Pull request by dbt**
Closes #1353.

Demo here:

http://jsfiddle.net/6py4rdr0/2/

### CODE REVIEW

File: src/core/formatters.ts [hunk 1]
```diff
@@ -131,6 +131,54 @@ export module Formatters {
...
+   *
+   * Numbers with a magnitude outside of (10^-precision, 10^(15+precision)) are shown using
+   * scientific notation to avoid creating extremely long decimal strings.  The inputs to the
+   * formatter are primarily expected to be ranges of common currency values.
+   *
+   * @param {number} [precision] the number of decimal places to show (default 3)
+   *
+   * @returns {Formatter} A formatter for large numbers.
+   *
+   *
```
Comment (aicioara) [inline src/core/formatters.ts]: Extra whitespace, delete the 2 blank lines

File: src/core/formatters.ts [hunk 2]
```diff
@@ -131,6 +131,54 @@ export module Formatters {
...
+   */
+  export function shortScale(precision = 3) {
+    verifyPrecision(precision);
+    var suffixes = "KMBTQ";
+    var efmt = d3.format("." + precision + "e");
+    var ffmt = d3.format("." + precision + "f");
```
Comment (aicioara) [inline src/core/formatters.ts]: I really have no idea what efmt and ffmt mean, so I guess it is not intuitive enough. Let's not abbreviate anything.

Comment (aicioara) [inline src/core/formatters.ts]: Ah, I got it now, fmt stands for formatter? Omitting vowels is not common in our codebase, so let's just call them fFormatter, eFormatter.

File: src/core/formatters.ts [hunk 3]
```diff
@@ -131,6 +131,54 @@ export module Formatters {
...
   /**
+   * Creates a formatter for values that displays abbreviated values
+   * and uses standard short scale suffixes (thousands, millions, billions, trillions,
+   * quadrillions).
+   *
+   * Numbers with a magnitude outside of (10^-precision, 10^(15+precision)) are shown using
+   * scientific notation to avoid creating extremely long decimal strings.  The inputs to the
+   * formatter are primarily expected to be ranges of common currency values.
+   *
+   * @param {number} [precision] the number of decimal places to show (default 3)
```
Comment (aicioara) [inline src/core/formatters.ts]: Those are not always decimal places. I believe for a precision of 3 you will show things like 123Q, 1234Q

File: src/core/formatters.ts [hunk 4]
```diff
@@ -131,6 +131,54 @@ export module Formatters {
...
+   *
+   */
+  export function shortScale(precision = 3) {
+    verifyPrecision(precision);
+    var suffixes = "KMBTQ";
+    var efmt = d3.format("." + precision + "e");
+    var ffmt = d3.format("." + precision + "f");
+    var max = Math.pow(10, (3 * suffixes.length) + precision + 1);
+    var min = Math.pow(10, -precision - 1);
+    return (d: any) => {
```
Comment (aicioara) [inline src/core/formatters.ts]: d: number?

File: src/core/formatters.ts (hunk 2 above)
Comment (aicioara) [inline src/core/formatters.ts]: what is an fFormatter btw?

File: test/core/formattersTests.ts [hunk 5]
```diff
@@ -168,6 +168,27 @@ describe("Formatters", () => {
...
+  describe("shortScale", () => {
+    it("shortens long numbers", () => {
+      var fmtr = Plottable.Formatters.shortScale();
+      assert.strictEqual(fmtr(1), "1.000", "small numbers format simply");
+      assert.strictEqual(fmtr(-1), "-1.000", "negative small numbers format simply");
+      assert.strictEqual(fmtr(2e3), "2.000K", "thousands numbers format using short scale magnitudes");
+      assert.strictEqual(fmtr(2e6), "2.000M", "millions numbers format using short scale magnitudes");
+      assert.strictEqual(fmtr(2e9), "2.000B", "billions numbers format using short scale magnitudes");
+      assert.strictEqual(fmtr(2e12), "2.000T", "trillions magnitude numbers format using short scale magnitudes");
+      assert.strictEqual(fmtr(2e15), "2.000Q", "quadrillions numbers format using short scale magnitudes");
```
Comment (aicioara) [inline test/core/formattersTests.ts]: Our max is 10^(3*5+3+1) = 10^19. 

Can we try 
- 10^19
- 10^19 - 1
- 10^18 ?

File: test/core/formattersTests.ts [hunk 6]
```diff
@@ -168,6 +168,27 @@ describe("Formatters", () => {
...
+      assert.strictEqual(fmtr(-2e15), "-2.000Q", "negative quadrillions numbers format using short scale magnitudes");
+      assert.strictEqual(fmtr(999.999), "999.999", "boundary cases are correct");
+      assert.strictEqual(fmtr(999.99999), "1.000K", "rounding cases are correct");
+      assert.strictEqual(fmtr(-999.99999), "-1.000K", "negative rounding cases are correct");
+      assert.strictEqual(fmtr(1000000), "1.000M", "boundary cases are correct");
+      assert.strictEqual(fmtr(-2e9), "-2.000B", "negative numbers format using short scale magnitudes");
```
Comment (aicioara) [inline test/core/formattersTests.ts]: Move this bellow the positive billion for consistency

File: test/core/formattersTests.ts [hunk 7]
```diff
@@ -168,6 +168,27 @@ describe("Formatters", () => {
...
+      assert.strictEqual(fmtr(1e37), "1.000e+37", "large magnitute number use scientific notation");
+      assert.strictEqual(fmtr(1e-7), "1.000e-7", "small magnitude numbers use scientific notation");
```
Comment (aicioara) [inline test/core/formattersTests.ts]: How about the edge case 10^(-3-1) = 10^-4 ?

File: test/core/formattersTests.ts [hunk 8]
```diff
@@ -168,6 +168,27 @@ describe("Formatters", () => {
...
+    });
+  });
+
```
Comment (aicioara) [inline test/core/formattersTests.ts]: Can we try a formatter with a specified precision (say 1)?

File: test/core/formattersTests.ts [hunk 9]
```diff
@@ -168,6 +168,27 @@ describe("Formatters", () => {
     });
   });
 
+  describe("shortScale", () => {
+    it("shortens long numbers", () => {
+      var fmtr = Plottable.Formatters.shortScale();
```
Comment (aicioara) [inline test/core/formattersTests.ts]: I prefer naming it `format`. Or anything that is not abbreviated by omitting voewels

File: test/core/formattersTests.ts (hunk 7 above)
Comment (aicioara) [inline test/core/formattersTests.ts]: Also (10^-4) - 1

File: src/core/formatters.ts (hunk 3 above)
Comment (jtlan) [inline src/core/formatters.ts]: I tried out the fiddle, and it looks like it always returns exactly `precision` decimal places.

**Comment by aicioara** [comment]
Taking this over.

File: plottable.d.ts [hunk 10]
```diff
@@ -661,6 +661,20 @@ declare module Plottable {
          */
         function siSuffix(precision?: number): (d: any) => string;
         /**
+         * Creates a formatter for values that displays abbreviated values
+         * and uses standard short scale suffixes (thousands, millions, billions, trillions,
+         * quadrillions).
+         *
+         * Numbers with a magnitude outside of (10 ^ (-precision), 10 ^ (15 + precision)) are shown using
+         * scientific notation to avoid creating extremely long decimal strings.  The inputs to the
+         * formatter are primarily expected to be ranges of common currency values.
```
Comment (bluong) [inline plottable.d.ts]: I don't really see the value of the last sentence here?

File: plottable.d.ts [hunk 11]
```diff
@@ -661,6 +661,20 @@ declare module Plottable {
...
+         *
+         * @param {number} [precision] the number of decimal places to show (default 3)
+         *
+         * @returns {Formatter} A formatter for large numbers.
```
Comment (bluong) [inline plottable.d.ts]: Unsure if the returns description is apt. A shortScale formatter returns a formatter with short scale formatting, not necessarily for large numbers?

File: src/core/formatters.ts [hunk 12]
```diff
@@ -131,6 +131,56 @@ export module Formatters {
...
+   *
+   * @param {number} [precision] the number of decimal places to show (default 3)
+   *
+   * @returns {Formatter} A formatter for large numbers.
+   */
+  export function shortScale(precision = 3) {
+    verifyPrecision(precision);
+    var suffixes = "KMBTQ";
+    var eFormatter = d3.format("." + precision + "e");
+    var fFormatter = d3.format("." + precision + "f");
```
Comment (bluong) [inline src/core/formatters.ts]: What are these `eFormatter`/`fFormatter` things?

Comment (bluong) [inline src/core/formatters.ts]: Let's use the terms in `d3.format

Comment (bluong) [inline src/core/formatters.ts]: Also instead of using `d3`'s `fixed` formatter, you could use our own formatter, not sure which one is better.

File: src/core/formatters.ts [hunk 13]
```diff
@@ -131,6 +131,56 @@ export module Formatters {
...
+        output = fFormatter(num);
+      } else {
+        output = fFormatter(num / factor) + suffixes[idx];
+      }
+      // catch rounding by the underlying d3 formatter
+      if ((num > 0 && output.substr(0, 5) === "1000.") || (num < 0 && output.substr(0, 6) === "-1000.")) {
+        if (idx < suffixes.length - 1) {
+          factor *= 1000;
+          idx += 1;
+          output = fFormatter(num / factor) + suffixes[idx];
```
Comment (bluong) [inline src/core/formatters.ts]: It seems that `factor` and `idx` seem to share some meaning together.  Instead of having 2 variables, why can't we have one variable (maybe `idx`) and then derive the other number where needed (factor seems to be consistently `Math.pow(1000, idx + 1)`

A lot of this code from my perspective is pretty confusing and having 2 variables that share similar information is part of the reason.

**Comment by bluong** [comment]
Documentation and cleanliness

**Comment by aicioara** [comment]
Agreed with all you said, but still using the d3 fixedFormatter. Amended

File: src/core/formatters.ts [hunk 14]
```diff
@@ -131,6 +131,51 @@ export module Formatters {
   }
 
   /**
+   * Creates a formatter for values that displays abbreviated values
+   * and uses standard short scale suffixes (thousands, millions, billions, trillions,
+   * quadrillions).
```
Comment (bluong) [inline src/core/formatters.ts]: Decently minor: Can you put the "thousands, millions, etc." part on a separate line?  I personally don't like that "quadrillions" just gets its own line.

Also, it'd be more clear that we only satisfy up to quadrillions.

**Comment by aicioara** [comment]
QE notes:

The way I tested this was through the console.
```javascript
> Plottable.Formatters.shortScale(3)(10)
< "10.000"
```
Where the first argument is the precision and the second argument is the formatted number.

Haven't tested this with axes though.

**Comment by crmorford** [comment]
Why does this have scientific notation for 0 instead of just 0.00? 
http://jsfiddle.net/6py4rdr0/3/
screen shot 2015-07-28 at 3 57 36 pm

**Comment by aicioara** [comment]
Good catch!

**Comment by crmorford** [comment]
One more thing: is it possible to get an error for non-whole-number input? http://jsfiddle.net/6py4rdr0/4/

**Comment by aicioara** [comment]
Sorry, I don't see what you mean. Mainly because it is random data.

**Comment by crmorford** [comment]
QE +1 pending on 
* errors or 
* math.floor() on non-whole-numbers passed in to Plottable.Formatters.shortScale()

**Comment by aicioara** [comment]
Created https://github.com/palantir/plottable/issues/2527 as well

**Comment by aicioara** [comment]
@crmorford We decided to Math.floor the wrong precision.
